File: BFS/10026/10026.cpp
#include <iostream>
#include <queue>
#include <cstring>
using namespace std;
int N, cnt = 0, cntForColorBlind = 0;
char adjList[101][101];
char adjListForColorBlind[101][101];
bool visited[101][101];
int dx[] = {-1, 0, 1, 0};
int dy[] = {0, 1, 0, -1};
typedef pair<int, int> pii;
void init() {
    cin.tie(nullptr);
    ios_base::sync_with_stdio(0);
}

void bfs(pii u) {
    queue<pii> q;
    q.push(u);
    while (!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        char color = adjList[x][y];
        q.pop();
        for (int idx = 0; idx < 4; idx++) {
            int nx = x + dx[idx];
            int ny = y + dy[idx];
            if (nx < 0 || ny < 0 || nx >= N || ny >= N) continue;
            if (!visited[nx][ny] && adjList[nx][ny] == color) {
                q.push(pii(nx, ny));
                visited[nx][ny] = true;
            }
        }
    }
}

void bfsForColorBlind(pii u) {
    queue<pii> q;
    q.push(u);
    while (!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        char color = adjListForColorBlind[x][y];
        q.pop();
        for (int idx = 0; idx < 4; idx++) {
            int nx = x + dx[idx];
            int ny = y + dy[idx];
            if (nx < 0 || ny < 0 || nx >= N || ny >= N) continue;
            if (!visited[nx][ny] && adjListForColorBlind[nx][ny] == color) {
                q.push(pii(nx, ny));
                visited[nx][ny] = true;
            }
        }
    }
}

int main() {
    init();
    cin >> N;
    for (int row = 0; row < N; row++) {
        for (int col = 0; col < N; col++) {
            cin >> adjList[row][col];
            adjListForColorBlind[row][col] = adjList[row][col] == 'G' ?
                'R' : adjList[row][col];
        }
    }
    for (int row = 0; row < N; row++) {
        for (int col = 0; col < N; col++) {
            if (!visited[row][col]) {
                visited[row][col] = true;
                bfs(pii(row, col));
                cnt++;
            }
        }
    }
    memset(visited, 0, sizeof(visited));
    for (int row = 0; row < N; row++) {
        for (int col = 0; col < N; col++) {
            if (!visited[row][col]) {
                visited[row][col] = true;
                bfsForColorBlind(pii(row, col));
                cntForColorBlind++;
            }
        }
    }
    cout << cnt << " " << cntForColorBlind << "\n";
    return 0;
}

File: BFS/10026/10026.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

private const val MAX = 100
private val areas = Array(MAX) { CharArray(MAX) }
private val visited = Array(MAX) { BooleanArray(MAX) { false } }
private val dx = listOf(-1, 0, 1, 0)
private val dy = listOf(0, 1, 0, -1)
private var n = 0

fun main(args: Array<String>) = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    n = readLine().toInt()

    repeat(n) { row ->
        val input = readLine()
        repeat(n) { col ->
            areas[row][col] = input[col]
        }
    }

    // normal
    var cnt = 0
    repeat(n) { row ->
        repeat(n) { col ->
            if (!visited[row][col]) {
                visited[row][col] = true
                bfs(row, col, false)
                cnt++
            }
        }
    }
    bw.write("$cnt ")

    // color blind
    resetVisited()
    cnt = 0
    repeat(n) { row ->
        repeat(n) { col ->
            if (!visited[row][col]) {
                visited[row][col] = true
                bfs(row, col, true)
                cnt++
            }
        }
    }
    bw.write("$cnt")

    bw.flush()
    bw.close()
}

private fun resetVisited() {
    repeat(n) { row ->
        repeat(n) { col ->
            visited[row][col] = false
        }
    }
}

private fun bfs(startY: Int, startX: Int, isColorBlind: Boolean) {
    val queue: Queue<Pair<Int, Int>> = LinkedList()
    queue.add(Pair(startY, startX))

    while (queue.isNotEmpty()) {
        val currY = queue.peek().first
        val currX = queue.peek().second
        queue.poll()

        repeat(4) {
            val dy = currY + dy[it]
            val dx = currX + dx[it]
            if (dy < 0 || dy >= n || dx < 0 || dx >= n) return@repeat

            if (areas[dy][dx] == areas[currY][currX] && !visited[dy][dx]) {
                visited[dy][dx] = true
                queue.add(Pair(dy, dx))
            } else if (isColorBlind && isRedAndGreen(areas[dy][dx], areas[currY][currX]) && !visited[dy][dx]) {
                visited[dy][dx] = true
                queue.add(Pair(dy, dx))
            }
        }
    }
}

fun isRedAndGreen(char1: Char, char2: Char) = setOf(char1, char2) == setOf('R', 'G')
